Initial batch of users of FSharp.Formatting. (#867)
fsprojects/FSharp.Formatting: docs/zero-to-hero.md

﻿---
category: Documentation
categoryindex: 1
index: 9
---
# From zero to hero: deploying to GitHub Pages

> This guide is meant for a typically setup for open-source projects on GitHub.  

Reviewer: ```suggestion
This guide is meant for a typical setup of open-source projects on GitHub.  
```

Both `build` and `watch` will generate the documentation for a solution and an input folder.  
The default folder for `--input` is the `./docs` folder, relative to the current working directory.

Typically your project will be structured like:

```
/repository-root
  YourSolution.sln
  ./docs
    index.md
    other-file.md
  ./src
    ./Project1/Project1.fsproj
    ./Project2/Project2.fsproj
```

It is recommended to have a single solution at the root. In some editors, it is more convenient to open a solution at the root, to easily manipulate any file in root repository folder.  
When users clone your repository locally, they cannot be confused on how they need to open the project in their IDE.

⚠️ Please avoid putting your solution in the `src` folder. When we open that solution, it can be more difficult to edit files in the `docs` folder, as we can sometimes only see the `src` folder.

That being said, let's create the `docs` folder and a first Markdown file named `index.md`.  
When `fsdocs` runs, it will transform this `index.md` file to `index.html`, which will be served at the root.

We can put `# Hello world` in the markdown file for now.

> Having this in play, should already serve the first page when we start the `watch` command:

Reviewer: ```suggestion
Having this in place, should already serve the first page when we start the `watch` command:
```

> dotnet fsdocs watch

Open [http://localhost:8901](http://localhost:8901) and you should see our first page!

🪄 You might notice that there are some images missing. You can add these in the `docs` folder in the right location.

## Generating API documentation

By default, `fsdocs` will generate API documentation for the configured `--projects`.  
When this flag is not specified, `fsdocs` will look for solutions or projects in the working directory.  
It will filter these found projects, the requirements are:

- Having `<OutputType>library</OutputType>`
- Having a binary, so you need to build your project first before the documentation can be generated.
- Not having `<IsTestProject>true</IsTestProject>`
- Having `<GenerateDocumentationFile>true</GenerateDocumentationFile>`

> 🪄 If you made some changes to in order to adhere to the rules, you may want to remove the `.fsdocs/cache` file.

Reviewer: ```suggestion
🪄 If you made some changes in order to adhere to the rules, you may want to remove the `.fsdocs/cache` file.
```

        <FsDocsLicenseLink>https://github.com/fsprojects/FSharp.AWS.DynamoDB/blob/master/License.md</FsDocsLicenseLink>
        <FsDocsReleaseNotesLink>https://github.com/fsprojects/FSharp.AWS.DynamoDB/blob/master/RELEASE_NOTES.md</FsDocsReleaseNotesLink>
        <PackageProjectUrl>https://fsprojects.github.io/FSharp.AWS.DynamoDB</PackageProjectUrl>
    </PropertyGroup>
</Project>
```

🪄 If you don't have any release notes yet, you could consider using [Ionide.KeepAChangelog](https://github.com/ionide/KeepAChangelog).

Running `dotnet fsdocs watch` will now yield:

```
  root --> https://github.com/fsprojects/FSharp.AWS.DynamoDB/
  ...
  fsdocs-license-link --> https://github.com/fsprojects/FSharp.AWS.DynamoDB/blob/master/License.md
  fsdocs-release-notes-link --> https://github.com/fsprojects/FSharp.AWS.DynamoDB/blob/master/RELEASE_NOTES.md
  ...
  fsdocs-repository-link --> https://github.com/fsprojects/FSharp.AWS.DynamoDB/
```

⚠️ Again, you might need to remove `.fsdocs/cache` in order for changes to be picked up! 

`<PackageProjectUrl>` is actually a very important property when you run `dotnet fsdocs build`.  

> `build` will generated static files for the targeted production environment. In our case, this will be GitHub Pages.

Reviewer: ```suggestion
`build` will generate static files for the targeted production environment. In our case, this will be GitHub Pages.
```

> Pages will host your files from https://github.com/user/project to on `https://user.github.io/project/` by default.  

Reviewer: ```suggestion
Pages will host your files from https://github.com/user/project on `https://user.github.io/project/` by default.  
```

You can change this by adding a custom domain so we need to be sure that all links and urls will be generated correctly during a build.

Let's now run `dotnet fsdocs build`.

`<PackageProjectUrl>` will replace the `{{root}}` substitution, which is used all over the place in the default template.  

⚠️ You want to ensure that the static files in the `output` folder (after running the build) have the correct links.

## Ignore generated files

> Alright, at this point we've made a lot of progress. I you are using `git` you want to add the following to your [.gitignore](https://git-scm.com/docs/gitignore) file.

Reviewer: ```suggestion
Alright, at this point we've made a lot of progress. If you are using `git` you want to add the following to your [.gitignore](https://git-scm.com/docs/gitignore) file.
```

# We need some permissions to publish to Github Pages
permissions:
  contents: write
  pages: write
  id-token: write

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      # Checkout the source code
      - uses: actions/checkout@v4
      # Setup dotnet, please use a global.json to ensure the right SDK is used!
      - name: Setup .NET
        uses: actions/setup-dotnet@v3
      # Restore the local tools
      - name: Restore tools
        run: dotnet tool restore
      # Build the code for the API documentation
      - name: Build code
        run: dotnet build -c Release YourSolution.sln
      # Generate the documentation files
      - name: Generate the documentation'
        run: dotnet fsdocs build --properties Configuration=Release
      # Upload the static files
      - name: Upload documentation
        uses: actions/upload-pages-artifact@v2
        with:
          path: ./output
  
  # GitHub Actions recommends deploying in a separate job.
  deploy:
    runs-on: ubuntu-latest
    needs: build
    steps:
      - name: Deploy to GitHub Pages
        id: deployment
        uses: actions/deploy-pages@v2
```

⚠️ Careful yaml is indentation sensitive!

## Next steps

> Mission accomplished right? If everything went well, you should have a published website at this point!  

Reviewer: ```suggestion
Mission accomplished, right? If everything went well, you should have a published website at this point!  
```
